apps/full-site-editing/full-site-editing-plugin/blocks/content-renderer/edit.js

```diff
@@ -0,0 +1,104 @@
+/* eslint-disable wpcalypso/jsx-classname-namespace */
+/**
+ * External dependencies
+ */
+import classNames from 'classnames';
+import { get } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { IconButton, Placeholder, Toolbar } from '@wordpress/components';
+import { withState } from '@wordpress/compose';
+import { BlockControls } from '@wordpress/editor';
+import { Fragment, RawHTML } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import PostAutocomplete from '../../components/post-autocomplete';
+import fetchPost from '../../lib/fetch-post';
+import './style.scss';
+
+const setSelectedPost = async ( attributes, setState ) => {
+	const { selectedPostId, selectedPostType } = attributes;
+	const selectedPost = await fetchPost( selectedPostId, selectedPostType );
+	setState( {
+		selectedPost,
+	} );
+};
+
+const ContentRendererEdit = withState( {
+	isEditing: false,
+	selectedPost: null,
+} )( ( { attributes, isEditing, selectedPost, setAttributes, setState } ) => {
+	const { align, selectedPostId } = attributes;
+
+	if ( !! selectedPostId && ! selectedPost ) {
+		setSelectedPost( attributes, setState );
+	}
+
+	const toggleEditing = () => setState( { isEditing: ! isEditing } );
+
+	const onSelectPost = post => {
+		setState( { isEditing: false, selectedPost: post } );
+		setAttributes( {
+			selectedPostId: get( post, 'id' ),
+			selectedPostType: get( post, 'type' ),
+		} );
+	};
+
+	const showToggleButton = ! isEditing || !! selectedPostId;
+	const showPlaceholder = isEditing || ! selectedPostId;
+	const showContent = ! isEditing && !! selectedPostId;
+
+	return (
+		<Fragment>
+			{ showToggleButton && (
+				<BlockControls>
+					<Toolbar>
+						<IconButton
+							className={ classNames( 'components-icon-button components-toolbar__control', {
+								'is-active': isEditing,
+							} ) }
+							label={ __( 'Change Content' ) }
+							onClick={ toggleEditing }
+							icon="edit"
+						/>
+					</Toolbar>
+				</BlockControls>
+			) }
+			<div
+				className={ classNames( 'a8c-content-renderer-block', {
+					[ `align${ align }` ]: align,
+				} ) }
+			>
+				{ showPlaceholder && (
+					<Placeholder
+						icon="layout"
+						label={ __( 'Content Renderer' ) }
+						instructions={ __( 'Select something to render' ) }
+					>
+						<div className="a8c-content-renderer-block__selector">
+							<PostAutocomplete
```

Comment:
Should we limit the results here to `wp_template` CPTs? I know this hasn't been merged yet but just wanted to check whether you think this is something that we should mark as a todo item for the future.

Reply:
I think that's something we might want to discuss more extensively.
On one hand, this block intent is to render layout elements.
On the other, what if I want to render a post inside a template?

A simple example that comes to mind would be a dynamic piece of content in the sidebar that wouldn't require me to update the entire template to change. (basically a widget of some sort).

Reply:
As per the next steps (I think we can tackle 3 out of 4 in follow ups, and the remaining in the other PR):

> - we should allow inserting both Content Preview and Content Renderer blocks only in wp_template CPT context. The former will serve as a placeholder post_content for the post that consumes the given template, and the second as a method for reusing template parts (e.g. header, footer) across multiple templates.

This makes sense, at least initially.
But then, I'm up for max flexibility, and so I'd be ok letting folks inserting a Renderer inside any post type.
(Thinking about it, Preview wouldn't make much sense outside a Template.

> - wp_template can contain only one Content Preview block (we'll permit rendering only one post within a template for now; no mixing of content and template parts).

Yes indeed! This is a 1-liner change: just add `multiple: false` to the block support properties. (I will actually do this in #32529 asap. No need to open an whole new PR for this).

> - Content Preview block will only show pages in its preview suggestions (or more broadly, not show wp_template CPTs).

Ok for not including Templates, but I'd leave it open to any kind of (publicly searchable) CPT.
Say, for a Woo product, or a JP portfolio page, etc.

> - Content Renderer block will show only wp_template CPTs in its post suggestions.

Ok, but see my [previous comment](https://github.com/Automattic/wp-calypso/pull/32581#discussion_r278968930). 🙂

Reply:
> A simple example that comes to mind would be a dynamic piece of content in the sidebar that wouldn't require me to update the entire template to change. (basically a widget of some sort).

Sounds like something that better fits a reusable block use-case?

Reply:
Yes, unless you also want to display it as a normal post. 😄 

I mean, folks use WP for the weirdest thing, and I'd love to keep giving them such freedom.
You're definitely right in scoping this down, just, let's not exclude this completely for future iterations.